### PR TITLE
Multithreaded scheduler

### PR DESCRIPTION
**Pull request by georgymh**
From the reverted pull request #3:

> Comments in code are hopefully pretty self-explanatory, but to summarize + extra details:
> 
> - Using Ray to run runners asynchronously
> - DMLRunnerActor is just a wrapper for DMLRunner (didn't want to mess with the tests for DMLRunner, which would clearly fail if I annotated DMLRunner as the actor)
> - Can run with as many runners as you want (default is 4), but ran into some weird errors when I was using too many with little speed boost. Feel free to change the default.
> - The test for speedup is fairly naive (Georgy mentioned that the scheduler is throwaway, so I didn't want to test by hardcoding stuff).
> - New to pull requests so I don't know if I'm the one merging...? If not, I'm fairly sure you can just add the new test file, add the new actor class, and replace the scheduler code with mine (unless there were significant changes to runner.py or smt

### CODE REVIEW

File: core/runner.py [hunk 1]
```diff
@@ -43,6 +43,7 @@ def run_job(self, job):
 
         If the runner is already executing a job, it silently does nothing.
         """
+
```
Comment (georgymh) [inline core/runner.py]: Can you remove this line.

File: core/runner_actor.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    DML Runner Actor
+
+    NOTE: This is a wrapper around DMLRunner so that testing on DMLRunner can be 
+    independent of interaction with Ray
+
+    NOTE 2: Please read https://ray.readthedocs.io/en/latest/actors.html. This 
+    class serves as an actor, where methods of each instance are run in parallel 
+    across instances. This is so that runners can run jobs independently of each
+    other.
+
```
Comment (georgymh) [inline core/runner_actor.py]: No need for the docs in lines 35-43. We wouldn't want to maintain docs in the original class and the actor.

File: core/runner_actor.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+# Add the parent directory to the PATH to allow imports.
```
Comment (georgymh) [inline core/runner_actor.py]: Why do we need lines 1-5? I think this is only useful for things inside `tests`. We're not running files directly and calling their main entrypoint anymore.

File: core/runner_actor.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import ray
+
+from core.runner import DMLRunner
+from custom.keras import model_from_serialized, get_optimizer
+from data.iterators import count_datapoints
+from data.iterators import create_train_dataset_iterator
+from data.iterators import create_test_dataset_iterator
+from core.utils.keras import train_keras_model, validate_keras_model
+from core.utils.keras import serialize_weights
+
```
Comment (georgymh) [inline core/runner_actor.py]: Extra line here. Our style guide is 2 lines between imports and code.

File: core/scheduler.py [hunk 5]
```diff
@@ -3,24 +3,24 @@
...
 from core.utils.dmljob import DMLJob
-from core.runner import DMLRunner
+from core.runner_actor import DMLRunnerActor
 
+
+ray.init()
 logging.basicConfig(level=logging.DEBUG,
                     format='[Scheduler] %(asctime)s %(levelname)s %(message)s')
 
 class DMLScheduler(object):
```
Comment (georgymh) [inline core/scheduler.py]: Write docstrings for each method. Some methods have comments right before their definition. These comments should be moved to the docstrings or inside the method definition.

File: tests/test_dmlscheduler.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+import numpy as np
+from custom.keras               import get_optimizer
+from core.runner                import DMLRunner
+from core.runner_actor          import DMLRunnerActor
+from models.keras_perceptron    import KerasPerceptron
+from core.utils.dmljob          import DMLJob, serialize_job, deserialize_job
+from core.scheduler             import DMLScheduler
+from examples.labelers          import mnist_labeler
+from queue import Queue
+print("testing")
```
Comment (georgymh) [inline tests/test_dmlscheduler.py]: Remove the printing statement here.

File: tests/test_dmlscheduler.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+
+import ray
+
+
+def make_dataset_path():
```
Comment (georgymh) [inline tests/test_dmlscheduler.py]: You can remove this function since you're not using it.

File: tests/test_dmlscheduler.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    return 'datasets/mnist'
+
+
+def make_config():
```
Comment (georgymh) [inline tests/test_dmlscheduler.py]: You can remove this function since you're not using it.

File: tests/test_dmlscheduler.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        scheduler._runners_run_next_jobs()
+    initial_weights = scheduler.processed.pop(0)
+    print(time.time()-start)
+    assert type(initial_weights) == list
+    assert type(initial_weights[0]) == np.ndarray
+    scheduler.reset()
+
+#Check that running 10 jobs with Ray takes at most half the 
+# time to run them serially (i.e. there is a speedup)
+def test_dmlscheduler_speedup_naive():
```
Comment (georgymh) [inline tests/test_dmlscheduler.py]: I don't think this is a good test. It's a good proof to justify why this scheduler is better, but I don't think it's something that should be run every time. It also depends on your hardware resources. I would remove this.

File: tests/test_dmlscheduler.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+def test_dmlscheduler_sanity():
+    model_json = make_model_json()
+    initialize_job = make_initialize_job(model_json)
+    scheduler.add_job(initialize_job)
+    scheduler._runners_run_next_jobs()
+    start = time.time()
+    while not scheduler.processed:
+        scheduler._runners_run_next_jobs()
+    initial_weights = scheduler.processed.pop(0)
+    print(time.time()-start)
```
Comment (georgymh) [inline tests/test_dmlscheduler.py]: Remove printing statement here. We're more interested in this not timing out rather than how long it takes in this stage (code).

File: tests/test_dmlscheduler.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    initialize_job = DMLJob(
+        "initialize",
+        model_json,
+        "keras"
+    )
+    return initialize_job
+
+scheduler = DMLScheduler()
+
+#Check that the scheduling/running functionality is maintained.
```
Comment (georgymh) [inline tests/test_dmlscheduler.py]: Put this comment inside as a docstring.

File: tests/test_dmlscheduler.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    start = time.time()
+    while len(scheduler.processed) < 10:
+        scheduler._runners_run_next_jobs()
+    fast = time.time() - start
+    print(slow)
+    print(fast)
+    assert fast*1.5 < slow
+    scheduler.reset()
+
+
```
Comment (georgymh) [inline tests/test_dmlscheduler.py]: Can we add a test that checks if 5 or 10 things can be scheduled at different times (for example with five things: scheduling 2 at the beginning, 2 after the 1st executes, one after the 4th executes) and verify that all jobs were executed? 

Also a test that checks if jobs are picked up and executed regardless of the scheduler being interacted with (basically that line 81 in the `scheduler.py` works)? (This would need to run fast, so the `run_frequency_in_mins` should be set to something like 5 seconds to keep the whole test under 10 seconds.)

Comment (georgymh) [inline tests/test_dmlscheduler.py]: Overall okay, but there are some changes to be made.

**Comment by georgymh** [comment]
@neeleshdodda44: Please see the comments above.

File: core/runner.py (hunk 1 above)
Comment (neeleshdodda44) [reply core/runner.py]: Done

File: core/runner_actor.py (hunk 2 above)
Comment (neeleshdodda44) [reply core/runner_actor.py]: Removed lines 35-43. The method comments are copied exactly from `DMLRunner`, are you sure you still want me to delete them?

File: core/runner_actor.py (hunk 3 above)
Comment (neeleshdodda44) [reply core/runner_actor.py]: Copied them exactly from `DMLRunner` and accidentally left them in. Removed.

File: core/runner_actor.py (hunk 4 above)
Comment (neeleshdodda44) [reply core/runner_actor.py]: Done.

File: core/scheduler.py (hunk 5 above)
Comment (neeleshdodda44) [reply core/scheduler.py]: Done.

File: tests/test_dmlscheduler.py (hunk 6 above)
Comment (neeleshdodda44) [reply tests/test_dmlscheduler.py]: Done.

File: tests/test_dmlscheduler.py (hunk 7 above)
Comment (neeleshdodda44) [reply tests/test_dmlscheduler.py]: Removed.

File: tests/test_dmlscheduler.py (hunk 8 above)
Comment (neeleshdodda44) [reply tests/test_dmlscheduler.py]: Removed.

File: tests/test_dmlscheduler.py (hunk 9 above)
Comment (neeleshdodda44) [reply tests/test_dmlscheduler.py]: I completely agree with your logic, and I actually thought about this problem for a while. Even though the test is naive (as I mentioned before), I still think it's necessary to have it. My reasoning is that if someone in the future messes with the class and accidentally breaks the threading and not the core functionality (i.e. every runner runs the same job), then we won't know since the tests would still pass.  I can always just lower the bar of the test if you're afraid of varying hardware resources causing occasional fails. 

If you have faith that the situation I described won't occur, though, I'll go ahead and delete the test.

File: tests/test_dmlscheduler.py (hunk 10 above)
Comment (neeleshdodda44) [reply tests/test_dmlscheduler.py]: Done

File: tests/test_dmlscheduler.py (hunk 11 above)
Comment (neeleshdodda44) [reply tests/test_dmlscheduler.py]: Done

File: tests/test_dmlscheduler.py (hunk 12 above)
Comment (neeleshdodda44) [reply tests/test_dmlscheduler.py]: The first test sounds good, I'll implement that.

As for the second test: since the CRON job method is something you wrote, I had a quick question. When the CRON job is set up, does it asynchronously call the method I specified every period (so the proper way to test would be to wait a few periods and then check if all jobs are done running)?

File: core/runner_actor.py [hunk 13]
```diff
@@ -24,14 +25,6 @@ class DMLRunnerActor(DMLRunner):
     """
     DML Runner Actor
 
-    NOTE: This is a wrapper around DMLRunner so that testing on DMLRunner can be
```
Comment (georgymh) [inline core/runner_actor.py]: I meant to delete the lines with "This class manages the training...This class only executes the jobs." which are now lines 28-35. The lines with the NOTEs were actually useful. Can you bring those lines back and delete 28-35?

Also, yes, even though you copied the method docstrings directly from DMLRunner, I believe we should remove them and just put:

```
"""See method docstring in the original class for more information."""
```

That way we do not maintain documentations in parallel across files.

File: core/scheduler.py [hunk 14]
```diff
@@ -57,17 +57,20 @@ def __init__(self, num_runners=4):
...
         self.queue.append(dml_job) 
 
     def create_runner(self):
+        """ Creates new runner. """
         return DMLRunnerActor.remote(self.dataset_path, self.runner_config)
 
-    #Check each job to see if it has a job running. If not, have the runner
-    # run the next job on the queue asynchronously of the others and collect 
-    # the result of the job that was running before (if applicable)
+    #
```
Comment (georgymh) [inline core/scheduler.py]: Remove this comment.

File: core/scheduler.py [hunk 15]
```diff
@@ -77,13 +80,15 @@ def _runners_run_next_jobs(self):
...
 
     def _start_cron(self, run_frequency_mins):
+        """ CRON job to run next jobs on runners, if applicable. """
         logging.info("Starting cron...")
         schedule.every(run_frequency_mins).minutes.do(self._runners_run_next_jobs)
         schedule.run_pending()
         logging.info("Cron started!")
 
-    #scheduler is a singleton, so need a way "start over"
+    #
```
Comment (georgymh) [inline core/scheduler.py]: Remove this comment.

File: core/scheduler.py [hunk 16]
```diff
@@ -77,13 +80,15 @@ def _runners_run_next_jobs(self):
...
     def reset(self):
+        """scheduler is a singleton, so need to get a new instance inplace """
         self.queue = deque()
         self.processed = []
```
Comment (georgymh) [inline core/scheduler.py]: Remove the commented code below this line.

File: core/scheduler.py [hunk 17]
```diff
@@ -57,17 +57,20 @@ def __init__(self, num_runners=4):
...
         return DMLRunnerActor.remote(self.dataset_path, self.runner_config)
 
-    #Check each job to see if it has a job running. If not, have the runner
-    # run the next job on the queue asynchronously of the others and collect 
-    # the result of the job that was running before (if applicable)
+    #
     def _runners_run_next_jobs(self):
+        """ Check each job to see if it has a job running. If not, have the runner run the 
+        next job on the queue asynchronously of the others and collect the result of the job
+        that was running before (if applicable) """
```
Comment (georgymh) [inline core/scheduler.py]: This may be small, but can you follow [PEP 257](https://www.python.org/dev/peps/pep-0257/) and use new lines instead of trailing spaces at the beginning and end of all multi-line docstrings?

File: tests/test_dmlscheduler.py [hunk 18]
```diff
@@ -1,7 +1,6 @@
 import pytest
```
Comment (georgymh) [inline tests/test_dmlscheduler.py]: There are still printing statements in this file to remove.

File: core/scheduler.py [hunk 19]
```diff
@@ -32,51 +32,66 @@ def getInstance():
             DMLScheduler()
         return DMLScheduler.__instance
 
-    def __init__(self):
+    #Default number of runners that are running at a time is 4
```
Comment (georgymh) [inline core/scheduler.py]: These kind of comments are better inside docstrings rather than outside methods. But I actually think we should remove it since the code is pretty much self-documenting at the function definition.

File: tests/test_dmlscheduler.py (hunk 12 above)
Comment (georgymh) [reply tests/test_dmlscheduler.py]: Ok. Once the first test is implemented and the minor things are changed we will be ready to merge.

For the second one, can you check [the documentation for the `schedule` module](http://schedule.readthedocs.io/en/stable/)? I found the module by googling but I wasn't able to test it until the multithreading capabilities were in place. I think you're correct on your approach to test it. We're not locked into the module, so if it's too hard to set up you can change it. We just want things to run as long as there are things to be run (as long as things are in the queue).

Comment (neeleshdodda44) [reply tests/test_dmlscheduler.py]: Completed both tests. 

Messaged you on Slack regarding the issues of `schedule` and `ray` altogether in the process of making the second test, and why I ended up using `multiprocessing`

File: core/runner_actor.py (hunk 13 above)
Comment (neeleshdodda44) [reply core/runner_actor.py]: Decided to move away from `ray`, so no need for the actor class anymore. 

Tl;dr resolved.

File: core/scheduler.py (hunk 14 above)
Comment (neeleshdodda44) [reply core/scheduler.py]: Removed.

File: core/scheduler.py (hunk 15 above)
Comment (neeleshdodda44) [reply core/scheduler.py]: Removed.

File: core/scheduler.py (hunk 16 above)
Comment (neeleshdodda44) [reply core/scheduler.py]: Removed.

File: core/scheduler.py (hunk 17 above)
Comment (neeleshdodda44) [reply core/scheduler.py]: Updated.

File: tests/test_dmlscheduler.py (hunk 18 above)
Comment (neeleshdodda44) [reply tests/test_dmlscheduler.py]: Removed (sorry I missed a few from debugging).

File: core/scheduler.py (hunk 19 above)
Comment (neeleshdodda44) [reply core/scheduler.py]: Removed.

**Comment by kiddyboots216** [comment]
@georgymh fyi i'm not approving PRs from now on so if this is good you should merge it
